src/test/java/edu/ucla/library/avpairtree/JobsQueueTest.java

```diff
@@ -0,0 +1,137 @@
+
+    private static final String ARK_REM = "ark:/21198/zz002dvxmn";
+
+    private static final String PATH_ROOT = "soul";
+
+    private static final String MAP = "jobs-queue";
+
+    /**
+     * Rule that creates the test context.
+     */
+    @Rule
+    public RunTestOnContext myContext = new RunTestOnContext();
+
+    /**
+     * Rule that provides access to the test method name.
+     */
+    @Rule
+    public TestName myNames = new TestName();
+
+    protected Logger getLogger() {
+        return LOGGER;
+    }
+
+    /**
+     * Test adding job to shared map
+     * @param aContext A test context
+     */
+    @Test
+    public void TestAddJobToQueue(final TestContext aContext) {
+        final CsvItem aItem = new CsvItem();
+        final Vertx vertx = myContext.vertx();
+        final Async asyncTask = aContext.async();
+
+        aItem.setItemARK(ARK_ADD);
+        aItem.setFilePath(SOUL_WAV);
+        aItem.setPathRoot(PATH_ROOT);
+
+        final JobsQueue JobsQueue = new JobsQueue(vertx);
+        final Future<Void> future = JobsQueue.addJobToQueue(aItem);
+
+        future.onSuccess(result -> {
+            vertx.sharedData().<String, CsvItem>getLocalAsyncMap(MAP, getMap -> {
+                if (getMap.succeeded()) {
+                    final AsyncMap<String, CsvItem> jobsQueue = getMap.result();
+                    jobsQueue.get(ARK_ADD).onSuccess(res -> {
+                        assertEquals(aItem, res);
+                        asyncTask.complete();
+                    });
+                } else {
+                    aContext.fail(getMap.cause());
+                }
+            });
+        }).onFailure(failure -> {
+            aContext.fail(failure);
+        });
+    }
+
+    /**
+     * Test removing jobs from shared map
+     * @param aContext A test context
+     */
+    @Test
+    public void TestRemoveJobInQueue(final TestContext aContext) {
+        final CsvItem aItem = new CsvItem();
+        final Async asyncTask = aContext.async();
+        final Vertx vertx = myContext.vertx();
+
+
+        aItem.setItemARK(ARK_REM);
+        aItem.setFilePath(SOUL_WAV);
+        aItem.setPathRoot(PATH_ROOT);
+
+        final JobsQueue JobsQueue = new JobsQueue(vertx);
+
+        //Put item in shared map
+        vertx.sharedData().<String, CsvItem>getLocalAsyncMap(MAP, getMap -> {
+            if (getMap.succeeded()) {
+                final AsyncMap<String, CsvItem> jobsMap = getMap.result();
+
+                jobsMap.put(ARK_REM, aItem, result -> {
+                    if (result.failed()) {
+                        aContext.fail(result.cause());
+                    }
+                });
+
+                final Future<Void> future = JobsQueue.removeJobInQueue(ARK_REM);
+                jobsMap.get(ARK_REM).onFailure(failure -> {
+                    aContext.fail();
+                }).onSuccess(res -> {
+                    assertEquals(res, null);
+                    asyncTask.complete();
+                });
```

Comment:
In this special case, these three async tasks (`jobsMap.put`, `JobsQueue.removeJobInQueue`, `jobsMap.get`) get coordinated properly, but that's just because they happen to use the nice and lawful Vert.x core asynchronous APIs all the way down. This will not be the case in general though (for example, if you have an async method that [runs some blocking I/O on a worker thread](https://vertx.io/docs/apidocs/io/vertx/core/Vertx.html#executeBlocking-io.vertx.core.Handler-) and then returns when that completes). So I'd argue it's always preferable to explicitly coordinate async stuff:

```suggestion
                jobsMap.put(ARK_REM, aItem, result -> {
                    if (result.failed()) {
                        aContext.fail(result.cause());
                    } else {
                        JobsQueue.removeJobInQueue(ARK_REM).compose(nil -> {
                            return jobsMap.get(ARK_REM);
                        }).onFailure(failure -> {
                            aContext.fail(failure.cause()); // Also, passing in the failure cause is a good idea
                        }).onSuccess(res -> {
                            assertEquals(null, res); // Also also, the expected value should always go first
                            asyncTask.complete();
                        });
                    });
                });
```

To be fair, this is really cumbersome since the void-returning async methods are all that's available with these older versions of Vert.x.

Reply:
Oh, I just realized this app uses Vert.x 4, in which case you could use the Future-returning versions if you wanted Javadocs in case you're interested:

- https://vertx.io/docs/4.2.7/apidocs/io/vertx/core/shareddata/SharedData.html#getLocalAsyncMap-java.lang.String-
- https://vertx.io/docs/4.2.7/apidocs/io/vertx/core/shareddata/AsyncMap.html#put-K-V-